lib/forwarder/server.go

```diff
@@ -0,0 +1,93 @@
+package forwarder
+
+import (
+	"context"
+	"net"
+	"tcplb/lib/core"
+)
+
+// CloseWriter represents something that can CloseWrite.
+//
+// Notable implementations in the standard library include:
+// - net.TCPCOnn
+// - tls.Conn
+type CloseWriter interface {
+	CloseWrite() error // CloseWrite shuts down the writer side of a connection.
+}
+
+type DuplexConn interface {
+	net.Conn
+	CloseWriter
+}
+
+// AuthenticatedConn represents an established authenticated
+// connection to a client.
+type AuthenticatedConn interface {
+	net.Conn
+	CloseWriter
+
+	// GetClientID attempts to extract the canonical ClientID representing
+	// the identity of an authenticated peer.
+	GetClientID() (core.ClientID, error)
+}
+
+// ClientReserver represents an entity that can limit "reservations"
+// by clients, as an abstraction of client rate limiting.
+//
+// Multiple goroutines may invoke methods on a ClientReserver
+// simultaneously.
+type ClientReserver interface {
+	// TryReserve attempts to acquire a reservation for the given client.
+	// If the attempt succeeds, nil is returned.
+	// If no reservations are available, the attempt returns an error.
+	// This call does not block.
+	TryReserve(ctx context.Context, c core.ClientID) error
+
+	// ReleaseReservation releases a reservation that was previously acquired
+	// for the given ClientID c by TryReserve.
+	ReleaseReservation(ctx context.Context, c core.ClientID) error
+}
+
+// Authorizer abstracts an authorization policy that
+// controls which clients are allowed to forward connections to which upstreams.
+//
+// Multiple goroutines may invoke methods on an Authorizer simultaneously.
+type Authorizer interface {
+	// AuthorizedUpstreams returns an UpstreamSet of upstreams that the ClientID c
+	// is authorized to access. If c is not authorized to access any upstreams,
+	// implementations should return an empty UpstreamSet and nil.
+	AuthorizedUpstreams(ctx context.Context, c core.ClientID) (core.UpstreamSet, error)
+}
+
+// BestUpstreamDialer dials the best upstream out of a set of candidates.
+//
+// Multiple goroutines may invoke methods on a BestUpstreamDialer simultaneously.
+type BestUpstreamDialer interface {
+	// DialBestUpstream considers the given candidate upstreams and attempts to connect to
+	// the "best" one (implementation defined). If successful, the winning Upstream is
+	// returned alongside a DuplexConn to that upstream, and nil error.
+	//
+	// If error is nil, the caller is responsible for closing the returned DuplexConn
+	// once finished with it to avoid leaking resources.
+	DialBestUpstream(ctx context.Context, candidates core.UpstreamSet) (core.Upstream, DuplexConn, error)
+}
+
+// Forwarder copies data between a client DuplexConn and an upstream DuplexConn.
+//
+// Multiple goroutines may invoke methods on a Forwarder simultaneously.
+type Forwarder interface {
```

Comment:
Similarly to the previous PR, we are following an interface declaration pattern that is unlike Go. Try to code and test against concrete types and push interfaces to the place they are used, you'll get a smaller and more idiomatic program as a result.

API methods with the same signature are not accidental in Go, they are usually done on purpose.